apps/system/utils/kdbg_heapinfo.c

```diff
@@ -43,42 +53,131 @@ extern size_t regionx_size[CONFIG_MM_REGIONS];
 extern int regionx_heap_idx[CONFIG_MM_REGIONS];
 #endif
 
+static int heapinfo_fd;
+static int heapinfo_display_flag = HEAPINFO_DISPLAY_ALL;
+static int heapinfo_mode = HEAPINFO_SIMPLE;
+static int heapinfo_pid = HEAPINFO_PID_ALL;
 #if CONFIG_MM_NHEAPS > 1
 extern heapinfo_total_info_t total_info;
 #endif
 
 #ifdef CONFIG_DEBUG_MM_HEAPINFO
-static void kdbg_heapinfo_init(FAR struct tcb_s *tcb, FAR void *arg)
+static void heapinfo_print_active_values(char *buf)
 {
+	int i;
+	stat_data stat_info[PROC_STAT_MAX];
 
-	tcb->peak_alloc_size = 0;
+	stat_info[0] = strtok(buf, " ");
 
-	printf("PID %d, peak allocated heap information is cleared\n", tcb->pid);
+	for (i = 1; i < PROC_STAT_MAX; i++) {
+		stat_info[i] = strtok(NULL, " ");
+	}
+	printf("%3s", stat_info[PROC_STAT_PID]);
+#if defined(CONFIG_SCHED_HAVE_PARENT) && !defined(HAVE_GROUP_MEMBERS)
+	printf(" | %5s", stat_info[PROC_STAT_PPID]);
+#endif
+	printf(" | %5s | %9s | %9s", stat_info[PROC_STAT_TOTALSTACK], stat_info[PROC_STAT_CURRHEAP], stat_info[PROC_STAT_PEAKHEAP]);
+#if CONFIG_TASK_NAME_SIZE > 0
+	printf(" | %s(", stat_info[PROC_STAT_NAME]);
+#else
+	printf("<noname>(");
+#endif
+	printf(")\n");
 }
 
-static void kdbg_heapinfo_task(FAR struct tcb_s *tcb, FAR void *arg)
+static int heapinfo_read_proc(pid_t pid)
 {
-	struct mm_allocnode_s *node;
-	node = (struct mm_allocnode_s *)(tcb->stack_alloc_ptr - SIZEOF_MM_ALLOCNODE);
+	int ret;
+	char *filepath;
+	char buf[HEAPINFO_BUFLEN];
 
-	printf("%3d", tcb->pid);
-#if defined(CONFIG_SCHED_HAVE_PARENT) && !defined(HAVE_GROUP_MEMBERS)
-	printf(" | %5d", tcb->ppid);
+	asprintf(&filepath, "%s/%d/%s", PROCFS_MOUNT_POINT, pid, "stat");
+	ret = kdbg_readfile(filepath, buf, HEAPINFO_BUFLEN, heapinfo_print_active_values);
+	free(filepath);
+	if (ret < 0) {
+		printf("Failed to read %s\n", filepath);
+		return ERROR;
+	}
+
+	return OK;
+}
+static void kdbg_heapinfo_set_option(char opt, int pid)
+{
+	switch (opt) {
+	case 'a':
+		heapinfo_mode = HEAPINFO_DETAIL_ALL;
+		heapinfo_display_flag = HEAPINFO_DISPLAY_ALL;
+		break;
+	case 'p':
+		heapinfo_mode = HEAPINFO_DETAIL_PID;
+		heapinfo_pid = pid;
+		heapinfo_display_flag = HEAPINFO_DISPLAY_ALL;
+		break;
+	case 'f':
+		heapinfo_mode = HEAPINFO_DETAIL_FREE;
+		heapinfo_display_flag = HEAPINFO_DISPLAY_ALL;
+		break;
+	case 'g':
+		heapinfo_mode = HEAPINFO_SIMPLE;
+		heapinfo_display_flag = HEAPINFO_DISPLAY_GROUP;
+		break;
+	case 'e':
+		heapinfo_mode = HEAPINFO_DETAIL_SPECIFIC_HEAP;
+		heapinfo_display_flag = HEAPINFO_DISPLAY_SPECIFIC_HEAP;
+		break;
+	default:
+		break;
+	}
+}
+
+static void kdbg_heapinfo_init(struct mm_heap_s *heap)
```

Comment:
this function clears values for the whole of heap?
I think 'heap' parameter is not needed because g_mmheap is passed.

Reply:
It is needed.